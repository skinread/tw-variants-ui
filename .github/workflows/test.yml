# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
