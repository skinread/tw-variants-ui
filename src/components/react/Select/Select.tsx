import { forwardRef, useRef } from 'react';
import { inputVariants, validityFromProps } from '@common';
import { type CommonFieldProps } from '@components/react';
import { FieldFeedback } from '../InputField/FieldFeedback';

type SelectPropsFiltered = Omit<React.ComponentPropsWithoutRef<'select'>, 'name' | 'localise'>;

export interface SelectProps extends CommonFieldProps, SelectPropsFiltered {
  /**
   * Array of labels and values as choices `{ label: "Option 1", value: "1"}`
   */
  options: {
    label: string;
    value: string;
  }[];
  /**
   * Place an optional zero-state entry in the input as a helpful text prompt
   */
  placeholder?: string;
  /**
   * Preselect an initial value in the select box
   */
  defaultValue?: SelectPropsFiltered['defaultValue'];
}

/**
 * Styled select box populated with an array. React `ref` enabled. Also exposes intrinsic `<select>` attributes.
 */
export const Select = forwardRef<HTMLSelectElement, SelectProps>(function VuiSelect(props, ref) {
  const {
    defaultValue = '',
    descriptionContent,
    feedbackContent,
    hasError,
    id: idProp,
    infoContent,
    label,
    options,
    placeholder,
    ...elemAttrs
  } = props;
  const {
    wrapper: wrapperStyle,
    label: labelStyle,
    labelText: labelTextStyle,
    input: inputStyle,
    descriptionWrapper: descriptionWrapperStyle,
  } = inputVariants({ type: 'select', validity: validityFromProps({ hasError }) });

  const id = useRef(idProp ?? `select-box-${props.name}`);
  const idFeedback = `${id.current}-feedback`;
  const idDescription = `${id.current}-description`;

  const Description = () => {
    if (!descriptionContent) return null;

    return (
      <div className={descriptionWrapperStyle()} id={idDescription}>
        {descriptionContent}
      </div>
    );
  };

  return (
    <div className={wrapperStyle()}>
      <label className={labelStyle()} htmlFor={id.current}>
        <span className={labelTextStyle()}>{label}</span>
        {infoContent && <span>{infoContent}</span>}
      </label>
      <select
        id={id.current}
        className={inputStyle()}
        defaultValue={defaultValue}
        aria-labelledby={feedbackContent ? idFeedback : undefined}
        aria-describedby={descriptionContent ? idDescription : undefined}
        {...elemAttrs}
        ref={ref}
      >
        {placeholder && (
          <option value="" disabled className="text-red-700">
            {placeholder}
          </option>
        )}
        {options.map((option, idx) => (
          <option key={idx} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      <FieldFeedback
        feedbackContent={feedbackContent}
        idFeedback={idFeedback}
        hasError={hasError}
      />
      <Description />
    </div>
  );
});
