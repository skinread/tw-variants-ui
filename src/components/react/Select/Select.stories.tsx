import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, within } from '@storybook/test';

import { Select } from './Select';

const meta = {
  title: 'React/Select',
  component: Select,
  tags: ['autodocs'],
  args: {
    label: 'Favourite fruit',
    options: [
      { label: 'Apple', value: 'a' },
      { label: 'Banana', value: 'b' },
      { label: 'Cherries', value: 'c' },
      { label: 'Dragon Fruit', value: 'd' },
      { label: 'Grapefruit', value: 'g' },
      { label: 'Kiwi', value: 'k' },
      { label: 'Mango', value: 'm' },
      { label: 'Orange', value: 'o' },
      { label: 'Raspberries', value: 'r' },
    ],
    name: 'demo-select-box',
    placeholder: 'Choose a fruit',
  },
  argTypes: {
    onChange: {
      action: 'changed',
    },
  },
} satisfies Meta<typeof Select>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {};

export const DefaultValue: Story = {
  args: {
    name: 'demo-select-box-default',
    defaultValue: 'js',
  },
};

export const Invalid: Story = {
  args: {
    name: 'demo-select-box-invalid',
    feedbackContent: 'Make sure you choose one',
    hasError: true,
  },
};

export const InteractionTest: Story = {
  args: {
    name: 'interaction-test',
  },
  play: async ({ args, canvasElement, step }) => {
    const canvas = within(canvasElement);
    const select = canvas.getByRole('combobox');
    const options = canvas.getAllByRole('option');

    await step('Initial state', async () => {
      await expect(options).toHaveLength(args.options.length + 1);
      await expect(select).toHaveTextContent(args.placeholder!);
    });

    await step('Selection', async () => {
      for (const idx of [2, 1, 0]) {
        await userEvent.selectOptions(select, args.options[idx]!.value);
        await expect(args.onChange).toHaveBeenCalled();
        // await expect((options[idx] as HTMLOptionElement).selected).toBe(true);
        await expect(select).toHaveValue(args.options[idx]?.value);
      }
    });

    // Testing Libary doesn't support keyboard navigation yet
    // ref https://github.com/testing-library/user-event/issues/786
    // await step('Keyboard navigation', async () => {
    //   await userEvent.click(select);
    //   await userEvent.keyboard('[ArrowDown][ArrowDown][Enter]');
    // });
  },
};
