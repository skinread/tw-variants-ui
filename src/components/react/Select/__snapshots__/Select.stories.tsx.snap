// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Select Basic smoke-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="select-box-demo-select-box"
  >
    <span class="label-text text-base font-bold">
      Favourite fruit
    </span>
  </label>
  <select id="select-box-demo-select-box"
          class="placeholder:text-subtle select select-primary text-base min-h-fit rounded-[--rounded-field] [&amp;:has(option[disabled]:default:checked)]:text-subtle h-[42px] px-3"
          name="demo-select-box"
  >
    <option value
            disabled
            class="text-red-700"
            selected
    >
      Choose a fruit
    </option>
    <option value="a">
      Apple
    </option>
    <option value="b">
      Banana
    </option>
    <option value="c">
      Cherries
    </option>
    <option value="d">
      Dragon Fruit
    </option>
    <option value="g">
      Grapefruit
    </option>
    <option value="k">
      Kiwi
    </option>
    <option value="m">
      Mango
    </option>
    <option value="o">
      Orange
    </option>
    <option value="r">
      Raspberries
    </option>
  </select>
</div>
`;

exports[`React/Select DefaultValue smoke-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="select-box-demo-select-box-default"
  >
    <span class="label-text text-base font-bold">
      Favourite fruit
    </span>
  </label>
  <select id="select-box-demo-select-box-default"
          class="placeholder:text-subtle select select-primary text-base min-h-fit rounded-[--rounded-field] [&amp;:has(option[disabled]:default:checked)]:text-subtle h-[42px] px-3"
          name="demo-select-box-default"
  >
    <option value
            disabled
            class="text-red-700"
    >
      Choose a fruit
    </option>
    <option value="a">
      Apple
    </option>
    <option value="b">
      Banana
    </option>
    <option value="c">
      Cherries
    </option>
    <option value="d">
      Dragon Fruit
    </option>
    <option value="g">
      Grapefruit
    </option>
    <option value="k">
      Kiwi
    </option>
    <option value="m">
      Mango
    </option>
    <option value="o">
      Orange
    </option>
    <option value="r">
      Raspberries
    </option>
  </select>
</div>
`;

exports[`React/Select InteractionTest play-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="select-box-interaction-test"
  >
    <span class="label-text text-base font-bold">
      Favourite fruit
    </span>
  </label>
  <select id="select-box-interaction-test"
          class="placeholder:text-subtle select select-primary text-base min-h-fit rounded-[--rounded-field] [&amp;:has(option[disabled]:default:checked)]:text-subtle h-[42px] px-3"
          name="interaction-test"
  >
    <option value
            disabled
            class="text-red-700"
            selected
    >
      Choose a fruit
    </option>
    <option value="a">
      Apple
    </option>
    <option value="b">
      Banana
    </option>
    <option value="c">
      Cherries
    </option>
    <option value="d">
      Dragon Fruit
    </option>
    <option value="g">
      Grapefruit
    </option>
    <option value="k">
      Kiwi
    </option>
    <option value="m">
      Mango
    </option>
    <option value="o">
      Orange
    </option>
    <option value="r">
      Raspberries
    </option>
  </select>
</div>
`;

exports[`React/Select Invalid smoke-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="select-box-demo-select-box-invalid"
  >
    <span class="label-text text-base font-bold">
      Favourite fruit
    </span>
  </label>
  <select id="select-box-demo-select-box-invalid"
          class="placeholder:text-subtle select select-primary text-base min-h-fit rounded-[--rounded-field] [&amp;:has(option[disabled]:default:checked)]:text-subtle border-2 h-[42px] px-3 select-error"
          aria-labelledby="select-box-demo-select-box-invalid-feedback"
          name="demo-select-box-invalid"
  >
    <option value
            disabled
            class="text-red-700"
            selected
    >
      Choose a fruit
    </option>
    <option value="a">
      Apple
    </option>
    <option value="b">
      Banana
    </option>
    <option value="c">
      Cherries
    </option>
    <option value="d">
      Dragon Fruit
    </option>
    <option value="g">
      Grapefruit
    </option>
    <option value="k">
      Kiwi
    </option>
    <option value="m">
      Mango
    </option>
    <option value="o">
      Orange
    </option>
    <option value="r">
      Raspberries
    </option>
  </select>
  <label class="leading-5 mt-2 [&amp;>label]:mt-0 text-error fill-error flex"
         id="select-box-demo-select-box-invalid-feedback"
  >
    <div class="-mt-0.5">
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           fill="currentColor"
           class="inline-block  w-6 h-6 mr-1.5 mb-0.5 icon-exclamation"
           aria-hidden="true"
      >
        <path d="m22.242,5.272l-3.515-3.515c-1.133-1.133-2.64-1.757-4.242-1.757h-4.971c-1.603,0-3.109.624-4.242,1.757l-3.515,3.514c-1.134,1.133-1.758,2.64-1.758,4.243v4.971c0,1.603.624,3.11,1.758,4.243l3.515,3.515c1.133,1.133,2.64,1.757,4.242,1.757h4.971c1.603,0,3.109-.624,4.242-1.757l3.515-3.514c1.134-1.133,1.758-2.64,1.758-4.243v-4.971c0-1.603-.624-3.11-1.758-4.243Zm-.242,9.213c0,1.069-.416,2.073-1.172,2.829l-3.515,3.515c-.756.755-1.76,1.171-2.828,1.171h-4.971c-1.068,0-2.072-.416-2.828-1.171l-3.515-3.515c-.756-.755-1.172-1.759-1.172-2.828v-4.971c0-1.069.416-2.073,1.172-2.829l3.515-3.515c.756-.755,1.76-1.171,2.828-1.171h4.971c1.068,0,2.072.416,2.828,1.171l3.515,3.515c.756.755,1.172,1.759,1.172,2.828v4.971Zm-11-.485V6c0-.552.447-1,1-1s1,.448,1,1v8c0,.552-.447,1-1,1s-1-.448-1-1Zm2.5,3.5c0,.828-.672,1.5-1.5,1.5s-1.5-.672-1.5-1.5.672-1.5,1.5-1.5,1.5.672,1.5,1.5Z"
              stroke="none"
        >
        </path>
      </svg>
    </div>
    Make sure you choose one
  </label>
</div>
`;
